cogs/Moderation.py

```diff
@@ -1,18 +1,114 @@
+    # and all subsequent calls will just return the cached values
+    async def get_attachment_hashes(self) -> set[Hash]:
+        if self.cached_attachment_hashes is None:
+            self.cached_attachment_hashes = set()
+
+            async with aiohttp.ClientSession() as session:
+                for attachment_url in self.attachment_urls[:5]:  # only load first 5 attachments to prevent abuse
+                    try:
+                        async with session.get(attachment_url) as resp:
+                            attachment_data = await resp.read()
+                    except Exception as e:
+                        # it's possible that Discord may have deleted the attachment, and so we would get a 404
+                        # furthermore, it's not super important that this function is 100% perfectly accurate
+                        # so it's fine to just log and ignore any errors
+                        log(
+                            f"Error occurred while downloading attachment for message fingerprinting. Attachment URL: `{attachment_url}`, Error: {e}"
+                        )
+                        continue
+
+                    self.cached_attachment_hashes.add(self.hash(attachment_data))
+
+        return self.cached_attachment_hashes
+
+    # returns True if two message fingerprints are similar (i.e. if the message is a 'multipost')
+    # specifically, if at least one of the attachments are identical or if the message body
+    # is identical and not blank
+    async def matches(self, other: "MessageFingerprint") -> bool:
+        # different authors cannot have matching fingerprints
+        if self.author_id != other.author_id:
+            return False
+
+        # messages in differing guilds are not applicable to fingerprinting
+        if self.guild_id != other.guild_id:
+            return False
+
+        # if there is content and it matches, then the fingerprint matches
+        if self.content_hash is not None and (self.content_hash == other.content_hash):
+            return True
+
+        # otherwise, at least one of the attachments must match
+        matching_attachments = await self.get_attachment_hashes() & await other.get_attachment_hashes()
+        return len(matching_attachments) > 0
 
 
-# TODO: Incomplete/in testing. Do not add to cog list.
 class Moderation(commands.Cog):
     def __init__(self, bot: commands.Bot) -> None:
         self.bot = bot
+        self.fingerprints: list[
+            MessageFingerprint
+        ] = []  # stores all user messages sent in the last minute (recent messages near the end)
+        self.delete_old_fingerprints_task.start()
 
-    @commands.slash_command(
-        name="mkick", description="Kicks multiple users from the server."
-    )
+    # temporarily commented out because this is not functional
+    """
+    @commands.slash_command(name="mkick", description="Kicks multiple users from the server.")
```

Comment:
tbh I'm not sure that a command like this would ever be used - seems a lot easier to just kick one-by-one. I'm not a moderator though so idk 🤷

Reply:
@GabeMillikan Yes, while it may not see many use cases, we primarily want to use it as a backup to our `/mban` command. We are frequently raided by a large number of spam bot accounts and `/mban` on the basis of join time.Often, there are innocent members who get caught in the crossfire, and having this `/mkick` functionality will provide them with a chance of rejoining. Again, this is mostly just a "backup."

If you could fix it, that would be awesome! Thank you

Reply:
Makes sense, I'll fix it up in another PR just so that the scope of this one doesn't get blown up too big :)